### PR TITLE
Break out S3 bucket options into standalone resources

### PR DESCRIPTION
**Pull request by mcdonnnj**
#  #

## 🗣 Description ##

This pull request breaks out AWS S3 bucket configuration parameters that received back-ported standalone resources with the [3.75.0 release of the Terraform AWS provider](https://github.com/hashicorp/terraform-provider-aws/releases/tag/v3.75.0). In addition resources have been added to block public access and enforce object ownership for the AWS S3 bucket configurations in this project. Lastly some cleanup was done on bucket name formation and how a production workspace is detected.

## 💭 Motivation and context ##

In `v3.75.0` of the Terraform AWS provider they back-ported the standalone S3 bucket resources to configure AWS S3 buckets that were introduced in v4 of the provider. This allows you to make changes to your Terraform configurations using v3 of the provider to prepare for the v4 provider. Making this change now helps prepare for updating to the v4 provider (and at this point prepare for the v5 provider).

## 🧪 Testing ##

Automated tests pass. Applying in both my workspace and the production workspace had the expected changes.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

File: versions.tf [hunk 1]
```diff
@@ -6,12 +6,12 @@ terraform {
   # major version currently being used.  This practice will help us
   # avoid unwelcome surprises.
   required_providers {
-    # Version 3.38.0 of the Terraform AWS provider is the first
-    # version to support default tags.
-    # https://www.hashicorp.com/blog/default-tags-in-the-terraform-aws-provider
+    # Version 3.75.0 of the Terraform AWS provider backports the changes to how
+    # AWS S3 resources are structured from the 4.0 release.
```
Comment (jsf9k) [inline versions.tf:10]: I would word this a little differently:
```suggestion
    # Version 3.75.0 of the Terraform AWS provider backports from the 4.0 release
    # the changes to how AWS S3 resources are structured.
```

File: exported_data_bucket.tf [hunk 2]
```diff
@@ -1,24 +1,56 @@
...
+  bucket = format("%s-%s", var.assessment_data_s3_bucket, local.production_workspace ? "production" : terraform.workspace)
 
   tags = {
     "Name" = "Exported Assessment Data"
   }
 
   lifecycle {
+    ignore_changes = [
+      # This should be removed when we upgrade the Terraform AWS provider to
+      # v4. It is necessary to use with the back-ported resources in v3.75 to
```
Comment (jsf9k) [inline exported_data_bucket.tf]: Consistency - we don't always hyphenate this word, even in the files in this PR:
```suggestion
      # v4. It is necessary to use with the backported resources in v3.75 to
```

File: lambda_bucket.tf [hunk 3]
```diff
@@ -1,28 +1,57 @@
...
+  bucket = format("%s-%s", var.assessment_data_import_lambda_s3_bucket, local.production_workspace ? "production" : terraform.workspace)
 
   tags = {
     "Name" = "Assessment Data Import Lambda"
   }
 
   lifecycle {
+    ignore_changes = [
+      # This should be removed when we upgrade the Terraform AWS provider to
+      # v4. It is necessary to use with the back-ported resources in v3.75 to
```
Comment (jsf9k) [inline lambda_bucket.tf]: Consistency - we don't always hyphenate this word, even in this PR:
```suggestion
      # v4. It is necessary to use with the backported resources in v3.75 to
```

**Review by jsf9k — APPROVED** [review]
I noticed a few very minor things, but nothing that stands in the way of my approval.

File: versions.tf (hunk 1 above)
Comment (mcdonnnj) [reply versions.tf:10]: Do we have any other alternatives? I am not a fan of how your suggestion reads.

Comment (jsf9k) [reply versions.tf:10]: It was just a suggestion.  You can ignore it.

Comment (mcdonnnj) [reply versions.tf:10]: Right, but if you aren't a fan of the language I'd rather come to a common ground on improving the language. @dav3r Do you have any thoughts?

Comment (mcdonnnj) [reply versions.tf:10]: I had the thought to see what's used in [cisagov/cyhy_amis](https://github.com/cisagov/cyhy_amis) to be consistent but it looks like I completely forgot that it's [where I got the language](https://github.com/cisagov/cyhy_amis/blob/5430bf305a75c0abe648cc8728077af57f0bf88a/terraform/versions.tf#L10-L12) in the first place. 😑

Comment (jsf9k) [reply versions.tf:10]: What is there is fine.  What you had makes sense, it just seemed a little awkward to me so I made a suggestion.

**Review by dav3r — APPROVED** [review]

**Comment by jsf9k** [comment]
I love this comment:

> Madness gripped me and I allowed a spellchecker to replace backported with back-ported, which is inconsistent with both this project and our general use of the term. Fortunately @jsf9k saved me from myself and we are rectifying this mistake.

As soon as I am done back-porting these changes, I shall pay a visit to the Burger-King and then continue on to Dulles.  I have an aero-plane to catch!

**Review by jsf9k — APPROVED** [review]
Approved with entirely too much enthusiasm!

**Review by jasonodoom — APPROVED** [review]

**Pull request merged**